### DISCUSSION TITLE
chp#7 Helper config..how to add multiple address in struct

### DISCUSSION THREAD

**Discussion by Abbasjafri-syed** [Cyfrin/foundry-full-course-cu#1186] [Q&A] [answered]
I am trying to add mulitple addresses at the same time in struct in the helpercofig section but it does not allow to do so as facing lot of issue at the same time...

```
  struct networkConfig {
        address[] priceFeed;
    }

    networkConfig[] networkConfigActive;

    constructor(address priceFeedAdd) {
        // if (block.chainid == 11155111) {
            // networkConfigActive = setUpSepoliaFeed(priceFeedAdd);
            networkConfigActive = setUpSepoliaFeed(1);
        // }
    }

    function setUpSepoliaFeed(uint priceFeedAdd) public pure returns (networkConfig memory) {
        address[] calldata dataFeeds = [
            "0x1b44F3514812d835EB1BDB0acB33d3fA3351Ee43",
            "0x694AA1769357215DE4FAC081bf1f309aDC325306",
            "0xc59E3633BAAC79493d908e63626716e204A45EdF",
            "0x1a81afB8146aeFfCFc5E50e8479e826E7D55b910",
            "0xA2F78ab2355fe2f984D808B5CeE7FD0A93D5270E"
        ];
        for (uint256 i; i < dataFeeds.length; i++) {
            // loop to run for all account
            networkConfigActive.priceFeed.push(networkConfig(dataFeeds[i])); // pushing account into array
        }
        return dataFeeds[priceFeedAdd];
    }
```

**Comment by Abbasjafri-syed** [comment]
@pacelliv can you please take a look at that

**Reply by ghost** [reply]
If you want to populate `networkActiveConfig` during construction, then move the logic to the constructor. 

There a few issues regarding types and how to initialize variables in `memory`.

1. In Solidity addresses are not written with quotes, because that turn them into the `string` type.
 
```diff
- "0x1b44F3514812d835EB1BDB0acB33d3fA3351Ee43"
+ 0x1b44F3514812d835EB1BDB0acB33d3fA3351Ee43
```

2. When you want to work with a dynamic type in memory (`bytes`, `string` or `array`), we don't do it as `calldata` instead we use the `memory` keyword.

`calldata` is a read-only temporary memory location that holds the arguments sent to the function by the caller (EOA or another smart contract). You cannot instantiate more `calldata` in the contract.

`memory` is a read-write temporary memory location that hold function arguments, local variables and **allow you to work with dynamic types**.

```diff
-        address[] calldata dataFeeds = [
+        address[5] memory dataFeeds = [
            // addresses
        ];
```

As shown above, by declaring `address[length]` this will create a fixed-array in memory. You cannot instantiate dynamic-size arrays like `address[]` in memory unless you do:

```javascript
address[] memory priceFeeds = new address[](length);
address[0] = priceFeed0
address[1] = priceFeed1
```

This show another way to intialize a fixed-size array in memory and the proper way to populate it. In memory it's forbidden to mutate arrays so methods like `push` and `pop` cannot be used, **rest assure this was not an issue in your code, but wanted to explain it**.

3. Populating `networkConfigActive`

```javascript
networkConfigActive.priceFeed.push(networkConfig(dataFeeds[i])); // pushing account into array
```

`networkConfigActive` is an array you cannot access the `priceFeed` member.

I wrote a lot so I understand if this causes confusion but you could use the following code to populate the config.

```javascript
  struct networkConfig {
        address[] priceFeed;
    }

    networkConfig[] networkConfigActive;

    constructor(address[] memory priceFeeds) {
        // initialize a new struct in memory
        networkConfig memory config;

        // initialize a new array in memory
        address[] memory newPriceFeeds = new address[](priceFeeds.length);

        // populate array
        for(uint256 i; i < priceFeeds.length; i++) {
            newPriceFeeds[i] = priceFeeds[i];
        }

        // set `priceFeed` member
        config.priceFeed = newPriceFeeds;
        // push new config to `networkConfigActive`
        networkConfigActive.push(config); // pushing account into array
    }
```

Finally, `struct`s are named using the PascalCase convention.

**Reply by Abbasjafri-syed** [reply]
thanks man....description also helps me clearing concept about memory and call data usage

**Comment by Abbasjafri-syed** [comment]
@pacelliv can you also please look at this...

i tried to get address which is stored at given index but unable to get that..
also please define it to have my concept cleared...

```
 function getUpSepoliaFeed(uint256 configIndex) external view returns (address) {

          networkConfig mdmofy config = networkConfigActive[configIndex];

         return config.priceFeed[configIndex];

}


    function setUpSepoliaFeed() public returns (networkConfig memory) {
        networkConfig memory config;

        // initialize a new array in memory
        address[] memory newPriceFeeds = new address[](5);
        newPriceFeeds[0] = 0x1b44F3514812d835EB1BDB0acB33d3fA3351Ee43;
        newPriceFeeds[1] = 0x694AA1769357215DE4FAC081bf1f309aDC325306;
        newPriceFeeds[2] = 0xc59E3633BAAC79493d908e63626716e204A45EdF;
        newPriceFeeds[3] = 0x1a81afB8146aeFfCFc5E50e8479e826E7D55b910;
        newPriceFeeds[4] = 0xA2F78ab2355fe2f984D808B5CeE7FD0A93D5270E;
            
        // set `priceFeed` member
        config.priceFeed =  newPriceFeeds;
        // push new config to `networkConfigActive`
        networkConfigActive.push(config); 
        return config;

        }
```

**Reply by ghost** [reply, marked as answer]
The way you have it, you can only pass index `0` to `getUpSepoliaFeed`, that's because you'd only pushed 1 config into `networkConfigActive`, so probably you need to pass two indexes. One to access a config and another index to access a price feed.

Update to something like this:

```diff
-  function getUpSepoliaFeed(uint256 configIndex) external view returns (address) {
+  function getUpSepoliaFeed(uint256 configIndex, uint256 priceFeedIndex) external view returns (address) {
```

Another thing you have some typo here:

```diff
- networkConfig mdmofy config = networkConfigActive[configIndex];
+ networkConfig memory config = networkConfigActive[configIndex];
```

**Reply by Abbasjafri-syed** [reply]
thanks for pointing out the issue.. this is how i get it

```
    function getUpSepoliaFeed(uint256 configIndex, uint256 priceFeedIndex) external view returns (address) {
          networkConfig memory config = networkConfigActive[configIndex];

         return config.priceFeed[priceFeedIndex];
}
```
